apollo-federation/src/sources/connect/models.rs

```diff
@@ -91,72 +105,122 @@ impl Connector {
+        let source_http = source.map(|s| &s.http);
+
+        let transport = HttpJsonTransport::from_directive(connect_http, source_http)?;
+
+        let parent_type_name = connect.position.field.type_name().clone();
+        let schema_def = &schema.schema_definition;
+        let on_query = schema_def
+            .query
+            .as_ref()
+            .map(|ty| ty.name == parent_type_name)
+            .unwrap_or(false);
+        let on_mutation = schema_def
+            .mutation
+            .as_ref()
+            .map(|ty| ty.name == parent_type_name)
+            .unwrap_or(false);
+        let on_root_type = on_query || on_mutation;
+
+        let id = ConnectId {
+            label: make_label(subgraph_name, &source_name, &transport),
+            subgraph_name: subgraph_name.to_string(),
+            source_name: source_name.clone(),
+            directive: connect.position,
+        };
 
-                    _ => None,
-                };
+        let entity_resolver = match (connect.entity, on_root_type) {
+            (true, _) => Some(EntityResolver::Explicit),
+            (_, false) => Some(EntityResolver::Implicit),
+            _ => None,
+        };
 
-                let request_variables = transport.variables().collect();
-                let response_variables = args.selection.external_variables().collect();
-
-                let connector = Connector {
-                    id: id.clone(),
-                    transport,
-                    selection: args.selection,
-                    entity_resolver,
-                    config: None,
-                    max_requests: None,
-                    spec,
-                    request_variables,
-                    response_variables,
-                };
+        let request_variables = transport.variables().collect();
+        let response_variables = connect.selection.external_variables().collect();
+
+        let connector = Connector {
+            id: id.clone(),
+            transport,
+            selection: connect.selection,
+            entity_resolver,
+            config: None,
+            max_requests: None,
+            spec,
+            request_variables,
+            response_variables,
+        };
 
-                Ok((id, connector))
-            })
-            .collect()
+        Ok((id, connector))
     }
 
     pub fn field_name(&self) -> &Name {
         self.id.directive.field.field_name()
     }
+
+    pub(crate) fn variable_references(
+        &self,
+    ) -> impl Iterator<Item = VariableReference<Namespace>> + '_ {
+        self.transport.variable_references().chain(
+            self.selection
+                .external_var_paths()
+                .into_iter()
+                .flat_map(PathSelection::variable_reference),
+        )
+    }
+
+    /// Create a field set for a `@key` using $args and $this variables.
+    pub(crate) fn resolvable_key(&self, schema: &Schema) -> Result<Option<Valid<FieldSet>>, ()> {
```

Comment:
Super small nit: it seems weird to use `Result` here and not have any data on the error case, especially if the caller of this method always maps the (empty) error into a custom error. Maybe consider making an enum for the three possible states?

Or this is overkill and just maybe consider adding some docs here as to what an empty error means in this context.

Reply:
ah yeah i was being lazy here. i should use some kind of internal error